### PR TITLE
rgbd: get rid of kernel build failure

### LINKED ISSUE #2830: rgbd: some kernels fail build
##### System information (version)

- OpenCV => recent master (ea41f89b40)
- Operating System / Platform => Ubuntu / ODROID-C4
- Compiler => GCC 9.3.0

##### Detailed description

```
[==========] Running 22 tests from 11 test cases.
[----------] Global test environment set-up.
[----------] 4 tests from KinectFusion
[ RUN      ] KinectFusion.lowDense
OpenCL program build log: rgbd/tsdf
Status -11: CL_BUILD_PROGRAM_FAILURE
-cl-mad-enable
<source>:1:9: error: unknown type name '__INT8_TYPE__'
typedef __INT8_TYPE__ int8_t;
        ^

<source>:63:20: warning: double precision constant requires cl_khr_fp64, casting to single precision
if(fabs(zStep.z) > 1e-5)
                   ^

error: Compiler frontend failed (error code 60)

unknown file: Failure
C++ exception with description "Failed to create kernel: <source>:1:9: error: unknown type name '__INT8_TYPE__'
typedef __INT8_TYPE__ int8_t;
        ^

<source>:63:20: warning: double precision constant requires cl_khr_fp64, casting to single precision
if(fabs(zStep.z) > 1e-5)
                   ^

error: Compiler frontend failed (error code 60)
" thrown in the test body.
[  FAILED  ] KinectFusion.lowDense (553 ms)
```

Removing this line satisfies the boards I own
https://github.com/opencv/opencv_contrib/blob/d733a80100bcc2145c2011ade96c53e7e2140767/modules/rgbd/src/opencl/tsdf.cl#L7

##### Steps to reproduce

Run `opencv_test_rgbd`

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by tomoaki0705**
Closes #2830 

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Linux OpenCL
```

### CODE REVIEW

**Comment by tomoaki0705** [comment]
Please hold to merge this PR.
I missed that even with this PR, kernel build still fails with following message

```
OpenCL program build log: rgbd/tsdf
Status -11: CL_BUILD_PROGRAM_FAILURE
-cl-mad-enable
<source>:1:9: error: unknown type name 'int8_t'; did you mean 'int8'?
typedef int8_t TsdfType;
        ^

<source>:10:1: error: unknown type name 'int8_t'; did you mean 'int8'?
int8_t res = (int8_t) ( (num * (-128)) );
^

<source>:10:23: error: called object type '<dependent type>' is not a function or function pointer
int8_t res = (int8_t) ( (num * (-128)) );
             ~~~~~~~~ ^

<source>:16:10: error: invalid conversion between vector type 'TsdfType' (aka 'int8') and scalar type 'float'
return ( (float) num ) / (-128);
         ^~~~~~~~~~~

error: Compiler frontend failed (error code 59)

unknown file: Failure
C++ exception with description "Failed to create kernel: <source>:1:9: error: unknown type name 'int8_t'; did you mean 'int8'?
typedef int8_t TsdfType;
        ^

<source>:10:1: error: unknown type name 'int8_t'; did you mean 'int8'?
int8_t res = (int8_t) ( (num * (-128)) );
^

<source>:10:23: error: called object type '<dependent type>' is not a function or function pointer
int8_t res = (int8_t) ( (num * (-128)) );
             ~~~~~~~~ ^

<source>:16:10: error: invalid conversion between vector type 'TsdfType' (aka 'int8') and scalar type 'float'
return ( (float) num ) / (-128);
         ^~~~~~~~~~~

error: Compiler frontend failed (error code 59)
" thrown in the test body.
[  FAILED  ] TSDF.valid_points (308 ms)
```

Even adding `typedef char int8_t`, then the OpenCL kernel call failed with mapping, so there's something heavy issue lies beneath this test failure.
Let me follow up.

**Comment by tomoaki0705** [comment]
Sorry for late update.
Now this PR is ready for merge
Before on ODROID-N2
```

[----------] 4 tests from TSDF
[ RUN      ] TSDF.raycast_normals
unknown file: Failure
C++ exception with description "Failed to create kernel: " thrown in the test body.
[  FAILED  ] TSDF.raycast_normals (159 ms)
[ RUN      ] TSDF.fetch_points_normals
unknown file: Failure
C++ exception with description "Failed to create kernel: " thrown in the test body.
[  FAILED  ] TSDF.fetch_points_normals (158 ms)
[ RUN      ] TSDF.fetch_normals
unknown file: Failure
C++ exception with description "Failed to create kernel: " thrown in the test body.
[  FAILED  ] TSDF.fetch_normals (157 ms)
[ RUN      ] TSDF.valid_points
unknown file: Failure
C++ exception with description "Failed to create kernel: " thrown in the test body.
[  FAILED  ] TSDF.valid_points (158 ms)
[----------] 4 tests from TSDF (632 ms total)
```
After
```
[----------] 4 tests from TSDF
[ RUN      ] TSDF.raycast_normals
[       OK ] TSDF.raycast_normals (247 ms)
[ RUN      ] TSDF.fetch_points_normals
[       OK ] TSDF.fetch_points_normals (1380 ms)
[ RUN      ] TSDF.fetch_normals
[       OK ] TSDF.fetch_normals (237 ms)
[ RUN      ] TSDF.valid_points
[       OK ] TSDF.valid_points (202 ms)
[----------] 4 tests from TSDF (2066 ms total)
```

Now, for [`fillPtsNrm`](https://github.com/opencv/opencv_contrib/blob/0b8aecd9ebb605f2c07c2cd21f987645eabcfd52/modules/rgbd/src/opencl/tsdf.cl#L703), [this kernel](https://github.com/opencv/opencv_contrib/blob/0b8aecd9ebb605f2c07c2cd21f987645eabcfd52/modules/rgbd/src/tsdf.cpp#L1163-L1164) doesn't have a fallback path to CPU (It directly raises an exception).
So on some devices, the test fails but it's no longer a kernel build failure.

ex.) on Firefly-RK3399
```
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from TSDF
[ RUN      ] TSDF.raycast_normals
[       OK ] TSDF.raycast_normals (937 ms)
[ RUN      ] TSDF.fetch_points_normals
OpenCL error CL_OUT_OF_RESOURCES (-5) during call: clEnqueueNDRangeKernel('fillPtsNrm', dims=3, globalsize=128x128x128, localsize=8x8x4) sync=true
unknown file: Failure
C++ exception with description "Failed to run kernel" thrown in the test body.
[  FAILED  ] TSDF.fetch_points_normals (1153 ms)
[ RUN      ] TSDF.fetch_normals
OpenCL error CL_OUT_OF_RESOURCES (-5) during call: clEnqueueNDRangeKernel('fillPtsNrm', dims=3, globalsize=128x128x128, localsize=8x8x4) sync=true
unknown file: Failure
C++ exception with description "Failed to run kernel" thrown in the test body.
[  FAILED  ] TSDF.fetch_normals (380 ms)
[ RUN      ] TSDF.valid_points
[       OK ] TSDF.valid_points (538 ms)
[----------] 4 tests from TSDF (3008 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (3012 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] TSDF.fetch_points_normals
[  FAILED  ] TSDF.fetch_normals
```

**Comment by tomoaki0705** [comment]
I want to emphasize that test of `rgbd` module was failing before this PR.
[4 tests are failing on Linux OpenCL builder](https://pullrequest.opencv.org/buildbot/builders/precommit-contrib_opencl_linux/builds/10139/steps/test_rgbd-ippicv-opencl/logs/stdio)


```
[==========] 22 tests from 11 test cases ran. (63875 ms total)
[  PASSED  ] 18 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] KinectFusion.lowDense
[  FAILED  ] KinectFusion.highDense
[  FAILED  ] KinectFusion.inequal
[  FAILED  ] KinectFusion.OCL
```

They are from `KinectFusion` test, which is irrelevant from this PR.
The tests I'm focusing is passing 

```
[----------] 4 tests from TSDF
[ RUN      ] TSDF.raycast_normals
[       OK ] TSDF.raycast_normals (1239 ms)
[ RUN      ] TSDF.fetch_points_normals
[       OK ] TSDF.fetch_points_normals (205 ms)
[ RUN      ] TSDF.fetch_normals
[       OK ] TSDF.fetch_normals (305 ms)
[ RUN      ] TSDF.valid_points
[       OK ] TSDF.valid_points (130 ms)
[----------] 4 tests from TSDF (1879 ms total)
```

So, the failure is not caused by this PR

**Review by savuor — APPROVED** [review]

**Comment by alalek** [comment]
@savuor 

Please take a look on "OpenCL Linux" builder failure:
- this patch: http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_opencl_linux/builds/10141
- another patch (almost master): http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_opencl_linux/builds/10140

**Comment by tomoaki0705** [comment]
I confirmed.
My modification does causes a test failure.
The failed test goes through the part that I modified.

Forgive me, this `KinectFusion` tests never passed on my local build farm.
It failed on every single Arm boards, so I believed that this was same on windows, too.
I tried on my old laptop and confirmed that before my modification, these tests pass but doesn't after my PR.

I don't want to fix entire problem (it's triggering another test one after another), so I''ll look for some meaningful portion of fix in a single PR.

Please hold this PR until I dig out the exact cause of this test failure.

File: modules/rgbd/src/tsdf.cpp [hunk 1]
```diff
@@ -859,14 +859,13 @@ static cv::UMat preCalculationPixNormGPU(int depth_rows, int depth_cols, Vec2f f
         throw std::runtime_error("Failed to create kernel: " + errorStr);
 
     AccessFlag af = ACCESS_READ;
-    UMat pixNorm = _pixNorm.getUMat(af);
```
Comment (alalek) [inline modules/rgbd/src/tsdf.cpp:862]: Agreed, `getUMat()` is local only wrapper - it should not be "returned" or stored somewhere.

**Comment by tomoaki0705** [comment]
Appreciate.
It took long, but I dig to the cause, at last.
I see no test failure on my Windows, Arm boards and on the official build farm.

Mainly 3 points to fix.
1. Kernel build failure: On Arm boards, `__INT8_TYPE__` was not predefined
2. Out of bounds access: Was happening on some arrays in the kernel
3. De-allocate before return: removed `getUMat()`

**Review by savuor — APPROVED** [review]

**Comment by tomoaki0705** [comment]
Anything else to be done before this PR gets merged?

**Review by alalek — APPROVED** [review]
Thank you!

**Pull request merged**